### PR TITLE
Ensure optimizer in backward works with 2d parallel

### PR DESCRIPTION
**Pull request by rohan-varma**
Summary: Test to ensure optimizer in backward works with 2D parallel.

Test Plan: CI

Differential Revision: D48508057

### CODE REVIEW

File: test/distributed/tensor/parallel/test_fsdp_2d_parallel.py [hunk 1]
```diff
@@ -215,21 +217,56 @@ def _test_2d_e2e_flow(
...
+                _apply_optimizer_in_backward(
+                    optimizer_class=torch.optim.Adam,
+                    params=param_group[0]["params"],
+                    optimizer_kwargs={"lr": param_group[0]["lr"]},
+                )
+                _apply_optimizer_in_backward(
+                    optimizer_class=torch.optim.Adam,
+                    params=param_group[1]["params"],
+                    optimizer_kwargs={"lr": param_group[1]["lr"]},
+                )
```
Comment (fduwjj) [inline test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:235]: is it possible to avoid duplicate code?

File: test/distributed/tensor/parallel/test_fsdp_2d_parallel.py [hunk 2]
```diff
@@ -215,21 +217,56 @@ def _test_2d_e2e_flow(
...
```
Comment (fduwjj) [inline test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:252]: ditto

File: test/distributed/tensor/parallel/test_fsdp_2d_parallel.py [hunk 3]
```diff
@@ -215,21 +217,56 @@ def _test_2d_e2e_flow(
...
+            else:
+                optim_2d = torch.optim.Adam(param_group, lr=0.01)
         else:
-            optim = torch.optim.Adam(model.parameters(), lr=0.01)
-            optim_2d = torch.optim.Adam(model_2d.parameters(), lr=0.01)
+            if optim_in_backward:
+                _apply_optimizer_in_backward(
```
Comment (fduwjj) [inline test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:257]: Does this also work with DDP? If so, do you mind also testing with DDP. It can be in a separate PR. Thanks!

File: test/distributed/tensor/parallel/test_fsdp_2d_parallel.py (hunk 1 above)
Comment (rohan-varma) [reply test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:235]: it can be deduped by 
```
for param_group_idx in len(param_group):
    _apply_optimizer_in_backward(
                    optimizer_class=torch.optim.Adam,
                    params=param_group[param_group_idx]["params"],
                    optimizer_kwargs={"lr": param_group[param_group_idx]["lr"]},
    )
```

but this doesn't add too much value IMO. lmk what you think

Comment (fduwjj) [reply test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:235]: Well, there are three places which are doing very similar thing. I do feel we can make it simpler here instead of copy paste six times.

File: test/distributed/tensor/parallel/test_fsdp_2d_parallel.py (hunk 3 above)
Comment (fegin) [reply test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:257]: `_apply_optimizer_in_backward` works for DDP too.

File: test/distributed/tensor/parallel/test_fsdp_2d_parallel.py (hunk 1 above)
Comment (awgu) [reply test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:235]: I agree that this usage of `_apply_optimizer_in_backward` is pretty verbose and ideally could be made less so.

**Review by awgu — APPROVED** [review]
Stamping to unblock -- @fduwjj feel free to do another review

File: test/distributed/tensor/parallel/test_fsdp_2d_parallel.py [hunk 4]
```diff
@@ -262,6 +300,13 @@ def test_2d_fsdp_integration_correctness_w_recompute_activation(self) -> None:
     def test_2d_fsdp_integration_use_orig_params(self) -> None:
         self._test_2d_e2e_flow(use_orig_params=True)
 
+    @with_comms
+    @skip_if_lt_x_gpu(4)
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:304]: yes, please use label `ciflow/periodic`

[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:304]: It's defined in .ci/pytorch/multigpu-test.sh

**Comment by fduwjj** [comment]
I don't have a strong opinion on this. If @rohan-varma wants to land this soon, I can clean it up as a follow-up PR.

**Review by fduwjj — APPROVED** [review]
LGTM

**Comment by rohan-varma** [comment]
macos tests are unrelated
